streaming/receiver.c

```diff
@@ -51,47 +51,235 @@ static void rrdpush_receiver_thread_cleanup(void *ptr) {
 
 #include "../collectors/plugins.d/pluginsd_parser.h"
 
-PARSER_RC streaming_timestamp(char **words, void *user, PLUGINSD_ACTION *plugins_action)
-{
-    UNUSED(plugins_action);
-    char *remote_time_txt = words[1];
-    time_t remote_time = 0;
-    RRDHOST *host = ((PARSER_USER_OBJECT *)user)->host;
-    struct plugind *cd = ((PARSER_USER_OBJECT *)user)->cd;
-    if (cd->version < VERSION_GAP_FILLING ) {
-        error("STREAM %s from %s: Child negotiated version %u but sent TIMESTAMP!", host->hostname, cd->cmd,
-               cd->version);
-        return PARSER_RC_OK;    // Ignore error and continue stream
+void send_replication_req(RRDHOST *host, char *st_id, time_t start, time_t end) {
+    char message[RRD_ID_LENGTH_MAX+60];
+    sprintf(message,"REPLICATE %s %ld %ld\n", st_id, start, end);
+    int ret;
+    debug(D_STREAM, "Replicate command: %s",message);
+#ifdef ENABLE_HTTPS
+    SSL *conn = host->stream_ssl.conn ;
+    if(conn && !host->stream_ssl.flags) {
+        ret = SSL_write(conn, message, strlen(message));
+    } else {
+        ret = send(host->receiver->fd, message, strlen(message), MSG_DONTWAIT);
     }
-    if (remote_time_txt && *remote_time_txt) {
-        remote_time = str2ull(remote_time_txt);
-        time_t now = now_realtime_sec(), prev = rrdhost_last_entry_t(host);
-        time_t gap = 0;
-        if (prev == 0)
-            info("STREAM %s from %s: Initial connection (no gap to check), remote=%ld local=%ld slew=%ld",
-                 host->hostname, cd->cmd, remote_time, now, now-remote_time);
-        else {
-            gap = now - prev;
-            info("STREAM %s from %s: Checking for gaps... remote=%ld local=%ld..%ld slew=%ld  %ld-sec gap",
-                 host->hostname, cd->cmd, remote_time, prev, now, remote_time - now, gap);
+#else
+    ret = send(host->receiver->fd, message, strlen(message), MSG_DONTWAIT);
+#endif
+    if (ret != (int)strlen(message))
+        error("Failed to send replication request for %s!", st_id);
```

Comment:
We have a `void` function that is responsible to send data and when it fails it writes a message inside `error.log`, but the fail can happens, because the connection was lost. Instead `void`, I think this function needs to return a value for the caller indicates to the thread that a reconnection is necessary.

Reply:
The replication requests are pumped by an external event. Each incoming point on the stream causes a replication request to be sent if the windows do not match. So although we do not explicitly return an error, there is a retry that is caused because the system has not reached the target state. At this point we do not know if a failure is transient or permanent so we rely on the logic in the main loop for the socket to handle errors / closures.